### DISCUSSION TITLE
Intellisense support for magic methods when using complex generic types

### DISCUSSION THREAD

**Discussion by girip11** [microsoft/pylance-release#4960] [Enhancement]
This pylance feature request is based on the discussion captured in the [issue](https://github.com/microsoft/pylance-release/issues/4938)

Consider the below snippet.

```python
# python 3.11 snippet
from typing import Any, Generic, Literal, TypeAlias, TypeVar

T = TypeVar("T", bound=str)

DS1FieldNames = Literal["col1", "col2", "col3"]

class GenericType(Generic[T]):
    def locate(self, value: T | list[T]) -> None:
        pass

    def __getattr__(self, value: T) -> Any:
        pass

    def __getitem__(self, value: T | list[T]) -> Any:
        pass

ConcreteType: TypeAlias = GenericType[DS1FieldNames]
c = ConcreteType()
```

Intellisense WORKS WELL when I type `c.locate("")` as well as `c.locate([""])`. Refer the image below.
![image](https://github.com/microsoft/pylance-release/assets/5471162/c057fb58-49ca-4daa-bb5d-fb268a198779)

Here are my feature requests with Intellisense support for the magic methods.

- Intellisense works with `__getattr__` when the type parameter is explicitly coded. But it currently doesn't support generic type parameters. **This will be very useful to restrict the possible attributes that can be called on a particular type.** 

![image](https://github.com/microsoft/pylance-release/assets/5471162/95b88396-e557-4b32-b440-40e066c6b5b5)

- And the magic method `__getitem__` currently supports generic type, but it doesn't support a signature like the one given in the above example. As discussed in the  issue #4938, I am adding this to the feature request.

- It will be great, in general if the Intellisense could work for other magic methods like `__setitem__`, `__setattr__` with generic types.

**Comment by debonte** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
